### PR TITLE
"Give NPC item" dialogue choice

### PR DESCRIPTION
**Pull request by Coolthulhu**
Note: contains a mechanics change that affects the player - removed the +30 encumbrance for being overloaded. It was required when overloading was a thing, now it all just falls out so that extra penalty only lasts for a turn.

New giving mechanic works like this:
- Only followers can be asked to wield/wear stuff
- NPC will evaluate own weapon first, counting own ammo to penalize guns with low ammo
- NPC will evaluate new item as a melee weapon, compare it to own weapon and wield it if it's better
- If the new item is a gun, NPC will count ammo for the gun in own inventory (and in the gun), evaluate the new item as a weapon (both melee and ranged) and wield it if it's better than own weapon
- If not wielded yet, NPC will try to put on the item. If it's not possible, they'll remove own clothing on the same body part first, then try to wear the new item
- Other items are NOT accepted yet! This means no drugs, bandages, ammo, mutagens etc. not even just to carry them

Currently the NPC will not explain why is the new item not OK, just say "Nope".

The last point can lead to easy robbery (just make NPC equip underwear until they lose carrying capacity), but I don't think it's much of a concern when you can just cause some sort of an accident to recover NPC's items.
For ease of design, I treated followers as minions player wants to preserve rather than acquaintances player might want to rob.

Also some refactoring and tweaks:
- Rewritten weapon-as-gun evaluation
- Split player encumbrance code into encumbrance from mutations/bionics and encumbrance from items. Currently not used, but will make my job easier when I update NPC armor evaluation
- Made NPCs print messages when putting on armor and removing it

### CODE REVIEW

**Comment by stk2008** [comment]
Wow so just to confirm we can now say use this gun insemtead of using the
debug menu

**Comment by Coolthulhu** [comment]
Yeah, but you may have problems giving the NPC ammo for said gun. And if the NPC doesn't have ammo in the inventory, the gun will have to be loaded to be seriously considered.

For now it will be useful mostly for melee and armor. Oh and for making that dude put down his damn flamethrower before he kills everybody.

**Comment by stk2008** [comment]
Lol yeah npcs with flamethrowers never end well tbh.

Thanks :)

**Comment by Rivet-the-Zombie** [comment]
Niiiiiiice!

**Comment by kevingranade** [comment]
Awesome, I agree that offering gifts to non followers should follow
different logic, and it's a good idea to just exclude it for now.

**Comment by kevingranade** [comment]
And the encumbrance penalty from being overloading still being there is an oversight, thanks for cleaning it up.

**Comment by Zireael07** [comment]
Wooow!

File: src/player.cpp [hunk 1]
```diff
@@ -10624,215 +10624,217 @@ bool player::wear_item(item *to_wear, bool interactive)
...
-    last_item = itype_id(to_wear->type->id);
-    worn.push_back(*to_wear);
+    last_item = itype_id(to_wear.type->id);
+    worn.push_back(to_wear);
 
-    if(interactive)
-    {
-        add_msg(_("You put on your %s."), to_wear->tname().c_str());
+    if( interactive ) {
+        add_msg( _("You put on your %s."), to_wear.tname().c_str() );
```
Comment (BevapDin) [inline src/player.cpp]: This still calls `add_msg`, is this intended? (Because all the other calls to it have been changed to `add_msg_if_player`.)

**Pull request merged**